include/insns/RV32I.h

```diff
@@ -142,67 +142,52 @@ class RV32I : public RevExt {
     return true;
   }
 
-  static constexpr auto& ebreak = nop;
+  static constexpr auto& ebreak    = nop;
 
   // Unimplemented Zicsr extension instructions
-  static constexpr auto& csrrw  = nop;
-  static constexpr auto& csrrs  = nop;
-  static constexpr auto& csrrc  = nop;
-  static constexpr auto& csrrwi = nop;
-  static constexpr auto& csrrsi = nop;
-  static constexpr auto& csrrci = nop;
+  static constexpr auto& csrrw     = nop;
+  static constexpr auto& csrrs     = nop;
+  static constexpr auto& csrrc     = nop;
+  static constexpr auto& csrrwi    = nop;
+  static constexpr auto& csrrsi    = nop;
+  static constexpr auto& csrrci    = nop;
 
   // Compressed instructions
 
-  // c.addi4spn %rd, $imm == addi %rd, x2, $imm
-  // if Inst.imm == 0; this is a HINT instruction and is effectively a NOP
-  static bool
-    caddi4spn( RevFeature* F, RevRegFile* R, RevMem* M, const RevInst& Inst ) {
-    return ( Inst.imm == 0 ? nop : addi )( F, R, M, Inst );
-  }
+  // c.addi4spn %rd, $imm == addi %rd, x2, $imm, $imm != 0
+  static constexpr auto& caddi4spn = addi;
```

Comment:
adding 0 and 0 and storing to x0 (which is always zero) is, of course, a harmless operation, however, are we losing the representation of `nop` here (and a few other places)?

Reply:
That particular diff is only a textual difference in the source code (not a semantic one), because in the new code, we do not need to disambiguate functionality at execution time -- the execute function can be fixed in the table entry.

`nop` is separated out as a distinct mnemonic here:
```c++
    { RevInstDefaults().SetMnemonic("addi %rd, %rs1, $imm" ).SetFunct3(   0b000).SetImplFunc(addi  ).Setrs2Class(RevRegClass::RegUNKNOWN).Setimm(FImm)                        .SetFormat(RVTypeI).SetOpcode(0b0010011).SetPredicate( []( uint32_t Inst ){ return DECODE_IMM12( Inst ) != 0; } ) },
    { RevInstDefaults().SetMnemonic("mv %rd, %rs1"         ).SetFunct3(   0b000).SetImplFunc(addi  ).Setrs2Class(RevRegClass::RegUNKNOWN).Setimm(FImm)                        .SetFormat(RVTypeI).SetOpcode(0b0010011).SetPredicate( []( uint32_t Inst ){ return DECODE_IMM12( Inst ) == 0 && ( DECODE_RS1( Inst ) != 0 || DECODE_RD( Inst ) != 0 ); } ) },
    { RevInstDefaults().SetMnemonic("nop"                  ).SetFunct3(   0b000).SetImplFunc(addi  ).Setrs2Class(RevRegClass::RegUNKNOWN).Setimm(FImm)                        .SetFormat(RVTypeI).SetOpcode(0b0010011).SetPredicate( []( uint32_t Inst ){ return DECODE_IMM12( Inst ) == 0 && ( DECODE_RS1( Inst ) == 0 && DECODE_RD( Inst ) == 0 ); } ) },

```
There are three separate cases handled here with predicates:
1. `addi` where the immediate != 0
2. `addi` where the immediate == 0 and one of the two registers is not `x0` -- this corresponds to a `mv` pseudoinstruction
3. `addi` where the immediate ==0 and both registers are `x0` -- this corresponds to the `nop` pseudoinstruction

All 3 cases are handled here, and the correct instruction is displayed.